app/auth.py

```diff
@@ -0,0 +1,239 @@
+        Gets the token(s) response from auth0, example:
+        
+        {   
+            'access_token': <str>
+            'id_token': <str>,
+            'scope': 'openid',
+            'expires_in': <int>.
+            'token_type': 'Bearer',
+            'expires_at': <int>
+        }
+        """
+
+        state = self.request.args.get("state", None)
+        if not state:
+            # TODO - handle this properly
+            raise Exception('Callback url should contain a "state" param')
+      
+        # Confirm auth server returned state matches initial
+        # state (guard against CSRF attacks)
+        # TODO - what do we do if it doesen't?        
+        assert state == self.get("state")
+
+        # Don't linger state once it's served its purpose
+        self.remove("state")
+
+        return self.client.fetch_token(
+            self.cfg.access_token_url,
+            authorization_response=self.request.url)
+
+
+    def logout(self):
+        """
+        Logs the user out via auth0 api.
+        
+        If that is successful, also removes the users token from
+        the cookie.
+        """
+
+        r = requests.get(self.cfg.logout_url, headers = {"client_id": self.cfg.client_id})
+        if r.status_code != 200:
+            self.logger.warning(f'Logout failed at {self.cfg.logout_url} with status code {r.status_code}')
+        else:
+            self.remove("access_token")
+
+
+    # TODO - in production this should be more generic, look for _something_
+    # in _some_ namespace. The equivient of _has_role() should just wrap that.
+    def _has_role(self, role) -> (bool):
+        """
+        Given the access token from the cookie, does the user in question have
+        the queried role under the self.cfg.roles_namespace key.
+        """
+
+        # Treat no token as does not have role, we don't want to be directing
+        # public users through a login.
+        decoded_cookie = self.decrypt_cookie()
+        if 'access_token' not in decoded_cookie:
+            return False
+
+        # Use token to get user information
+        headers = {'Authorization': f'Bearer {decoded_cookie["access_token"]}'}
+        r = requests.post(self.cfg.userinfo_url, headers=headers)
+        if not r.ok:
+            # TODO - handle this properly
+            raise NotImplementedError('Logout url not working')
+        self.logger.debug(f'Logout got status code {r.status_code}')
+        
+        user_info = r.json()
+
+        # User has authenticated, but has no assigned roles
+        if self.cfg.roles_namespace not in user_info:
+            return False
+
+        return role in user_info[self.cfg.roles_namespace]
+
+
+    def has_admin(self) -> (bool):
+        """
+        Does the current user cookie contain an access token that
+        identifies them as having then "admin" role under the
+        roles_namespace key.
+        """
+        return self._has_role("admin")
+
+
+    def pristine(self):
```

Comment:
pep8 - too many blank lines between functions

Reply:
indeed, let's use [black](https://pypi.org/project/black/) on this project though.